docs/csharp/language-reference/statements/iteration-statements.md

```diff
@@ -0,0 +1,163 @@
+
+:::code language="csharp" interactive="try-dotnet-method" source="snippets/iteration-statements/ForStatement.cs" id="MultipleExpressions":::
+
+All the sections of the `for` statement are optional. For example, the following code defines the infinite `for` loop:
+
+:::code language="csharp" source="snippets/iteration-statements/ForStatement.cs" id="InfiniteLoop":::
+
+## The `foreach` statement
+
+The `foreach` statement executes a statement or a block of statements for each element in an instance of the type that implements the <xref:System.Collections.IEnumerable?displayProperty=nameWithType> or <xref:System.Collections.Generic.IEnumerable%601?displayProperty=nameWithType> interface, as the following example shows:
+
+:::code language="csharp" interactive="try-dotnet-method" source="snippets/iteration-statements/ForeachStatement.cs" id="WithIEnumerable":::
+
+The `foreach` statement isn't limited to those types. You can use it with an instance of any type that satisfies the following conditions:
+
+- A type has the public parameterless `GetEnumerator` method. Beginning with C# 9.0, the `GetEnumerator` method can be a type's [extension method](../../programming-guide/classes-and-structs/extension-methods.md).
+- The return type of the `GetEnumerator` method has the public `Current` property and the public parameterless `MoveNext` method whose return type is `bool`.
+
+The following example uses the `foreach` statement with an instance of the <xref:System.Span%601?displayProperty=nameWithType> type, which doesn't implement any interfaces:
+
+:::code language="csharp" source="snippets/iteration-statements/ForeachStatement.cs" id="WithSpan" :::
+
+Beginning with C# 7.3, if the enumerator's `Current` property returns a [reference return value](../keywords/ref.md#reference-return-values) (`ref T` where `T` is the type of a collection element), you can declare an iteration variable with the `ref` or `ref readonly` modifier, as the following example shows:
+
+:::code language="csharp" source="snippets/iteration-statements/ForeachStatement.cs" id="RefIterationVariable" :::
+
+If the `foreach` statement is applied to `null`, a <xref:System.NullReferenceException> is thrown. If the source collection of the `foreach` statement is empty, the body of the `foreach` statement isn't executed and skipped.
+
+### await foreach
+
+Beginning with C# 8.0, you can use the `await foreach` statement to consume an asynchronous stream of data, that is, the collection type that implements the <xref:System.Collections.Generic.IAsyncEnumerable%601> interface. Each iteration of the loop may be suspended while the next element is retrieved asynchronously. The following example shows how to use the `await foreach` statement:
```

Comment:
Is it worth mentioning `IAsyncEnumerator`, and its relationship with `IEnumerator`: meaning a `Current` property and a `MoveNextAsync` method that returns a `Task<bool>`?

I think yes, but that's a very weakly held opinion. Thoughts @pkulikov

Reply:
@BillWagner do you mean we should mention the pattern-based compilation as we do for a synchronous foreach? I think that would be useful to be consistent with the beginning of the section

Reply:
@pkulikov  Yes, that's what I mean. Thanks.

Reply:
@BillWagner I've updated the article; please check the latest commit.

Reply:
That's perfect @pkulikov 

I'll :shipit: as soon as the build finishes.

Thanks again!

Reply:
@BillWagner thank you for the review and the suggestion for `await foreach`